src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs

```diff
@@ -795,22 +798,27 @@ public Task<List<RawFunctionMetadata>> GetFunctionMetadata()
 
         internal Task<List<RawFunctionMetadata>> SendFunctionMetadataRequest()
         {
-            // reset indexing task when in case we need to send another request
-            _functionsIndexingTask = new TaskCompletionSource<List<RawFunctionMetadata>>(TaskCreationOptions.RunContinuationsAsynchronously);
+            _workerChannelLogger.LogDebug("Fetching worker metadata, FunctionMetadataReceived set to: {functionMetadataReceived}", _functionMetadataRequestSent);
 
-            RegisterCallbackForNextGrpcMessage(MsgType.FunctionMetadataResponse, _functionLoadTimeout, 1,
-                msg => ProcessFunctionMetadataResponses(msg.Message.FunctionMetadataResponse), HandleWorkerMetadataRequestError);
+            if (!_functionMetadataRequestSent)
```

Comment:
Curious to know what use case might be causing 2 threads to invoke `SendFunctionMetadataRequest` method simultaneously.

Reply:
Also, this approach here is not really thread safe since the boolean variable is declared outside of the method scope. If 2 threads are calling this method simultaneously, it is possible that the code between 805 and 817 could be executed more than once (Imagine thread 1 called this method & started executing the code (805-817), but has not set the boolean value to `true` yet and thread 2 called this method).

This is why I am curious to know what scenarios is causing this issue.

Reply:
There isn't a scenario where we should see 2 threads executing this at once. 
There was a bug in the hot reload scenario where we would reuse the same language worker between script host restarts.  If a user makes changes to function files in quick successions we would have multiple host restarts. That is where we would see multiple threads calling this at the same time as we were not shutting down the channel. 

The change in the PR below fixes it. However, i am adding this in case we have an unknown scenario calling this multiple times, since we saw customers, app running into that scenario.
https://github.com/Azure/azure-functions-host/pull/9244/files

Reply:
@kshyju have updated the code to be thread safe.